### PR TITLE
Porting Button from .NET nanoFramework

### PR DESCRIPTION
**Pull request by zoekdestep**
This is a port from the Button binding from the .NET nanoFramework (link to [binding](https://github.com/nanoframework/nanoFramework.IoT.Device/tree/develop/devices/Button)).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/iot/pull/1680)

### CODE REVIEW

File: src/devices/Button/ButtonBase.cs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+        private long _lastPress = DateTime.MinValue.Ticks;
+        private Timer? _holdingTimer;
+
+        /// <summary>
+        /// Delegate for button pressed.
+        /// </summary>
+        /// <param name="sender">The sender object.</param>
+        /// <param name="e">The event data.</param>
+        public delegate void ButtonPressedDelegate(object sender, EventArgs e);
```
Comment (pgrawehr) [inline src/devices/Button/ButtonBase.cs:33]: No need to declare these manually in .Net. Use `Action<T>`

File: src/devices/Button/ButtonHoldingState.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Iot.Device.Button
+{
+    /// <summary>
+    /// The different button holding states.
+    /// </summary>
+    public enum ButtonHoldingState
+    {
+        /// <summary>Started</summary>
```
Comment (pgrawehr) [inline src/devices/Button/ButtonHoldingState.cs:11]: Documentation could be improved. The meaning of this enum is not obvious

File: src/devices/Button/GpioButton.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        /// <summary>
+        /// Initialization of the button.
+        /// </summary>
+        /// <param name="buttonPin">GPIO pin of the button.</param>
+        /// <param name="pullUp">If the system is pullup (false = pulldown).</param>
+        /// <param name="doublePress">Max ticks between button presses to count as doublepress.</param>
+        /// <param name="holding">Min ms a button is pressed to count as holding.</param>
+        /// <param name="gpio">Gpio Controller.</param>
+        /// <param name="shouldDispose">True to dispose the GpioController.</param>
+        public GpioButton(int buttonPin, bool pullUp, TimeSpan doublePress, TimeSpan holding, GpioController? gpio = null, bool shouldDispose = true)
```
Comment (pgrawehr) [inline src/devices/Button/GpioButton.cs:43]: I think `pullUp` should be an enum, because we might want to support internal pull-ups/pull-downs as well.

File: src/devices/Button/GpioButton.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                    }
+
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Cleanup.
+        /// </summary>
+        public new void Dispose()
```
Comment (pgrawehr) [inline src/devices/Button/GpioButton.cs:92]: Why is this required?

File: src/devices/Button/GpioButton.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                    _gpioController = null!;
+                }
+                else
+                {
+                    _gpioController.UnregisterCallbackForPinValueChangedEvent(_buttonPin, PinStateChanged);
+                    _gpioController.ClosePin(_buttonPin);
+                }
+
+                base.Dispose(disposing);
+                _disposed = true;
```
Comment (pgrawehr) [inline src/devices/Button/GpioButton.cs:122]: Put base call outside of if.

File: src/devices/Button/GpioButton.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            if (disposing)
+            {
+                if (_shouldDispose)
+                {
+                    _gpioController?.Dispose();
+                    _gpioController = null!;
+                }
+                else
+                {
+                    _gpioController.UnregisterCallbackForPinValueChangedEvent(_buttonPin, PinStateChanged);
```
Comment (pgrawehr) [inline src/devices/Button/GpioButton.cs:117]: This should be done in either case.

File: src/devices/Button/GpioButton.cs (hunk 5 above)
Comment (zoekdestep) [reply src/devices/Button/GpioButton.cs]: Thank you! Should be fixed. Will resolve for now - if you feel like the fix doesn't answer your comment, feel free to reopen the conversation.

File: src/devices/Button/GpioButton.cs (hunk 6 above)
Comment (pgrawehr) [reply src/devices/Button/GpioButton.cs:117]: @Ellerbach It probably is, but still for consistency, I would say we should unregister before disposing.

File: src/devices/Button/ButtonBase.cs (hunk 1 above)
Comment (pgrawehr) [reply src/devices/Button/ButtonBase.cs]: @zoekdestep Thanks. In such cases, you can just silently resolve a conversation. This is considered equivalent to "I think I have addressed the issue". Less typing ;-)
